api/envoy/service/network_ext_proc/v3/network_external_processor.proto

```diff
@@ -0,0 +1,183 @@
+// * Control connection lifecycle (continue, close gracefully, or reset)
+//
+// Use cases include:
+// * Custom protocol inspection and modification
+// * Advanced traffic manipulation
+// * Security scanning and filtering
+// * Dynamic connection management
+//
+// The service uses a bidirectional gRPC stream, maintaining state throughout
+// the connection lifetime while allowing asynchronous processing.
+
+// NetworkExternalProcessor service defines the contract between Envoy and
+// external processing servers for L4 network traffic.
+service NetworkExternalProcessor {
+  // Process establishes a bidirectional stream between Envoy and the external
+  // processing server. Envoy sends ProcessingRequests containing network data
+  // and the server responds with ProcessingResponses containing processing
+  // decisions and potentially modified data.
+  //
+  // The server should handle processing timeout properly to avoid blocking
+  // network traffic for extended periods. Any uncaught exceptions will
+  // be treated as errors and will terminate the stream.
+  //
+  // Implementation note: The server should process requests in the order
+  // they are received to maintain proper sequencing of network traffic.
+  rpc Process(stream ProcessingRequest) returns (stream ProcessingResponse) {
+  }
+}
+
+// ProcessingRequest contains data sent from Envoy to the external processing server.
+// Each request contains either read data (from client) or write data (to client)
+// along with optional metadata.
+message ProcessingRequest {
+  oneof request {
+    option (validate.required) = true;
+
+    // ReadData contains the network data intercepted in the request path (client to server).
+    // This is sent to the external processor when data arrives from the downstream client.
+    ReadData read_data = 1;
+
+    // WriteData contains the network data intercepted in the response path (server to client).
+    // This is sent to the external processor when data arrives from the upstream server.
+    WriteData write_data = 2;
+  }
+
+  // Optional metadata associated with the request.
+  // This can include connection properties, filter configuration, and any other
+  // contextual information that might be useful for processing decisions.
+  //
+  // The metadata is not automatically propagated from request to response.
+  // The external processor must include any needed metadata in its response.
+  config.core.v3.Metadata metadata = 3;
+}
+
+// ProcessingResponse contains the response from the external processing server to Envoy.
+// Each response corresponds to a ProcessingRequest and indicates how the network
+// traffic should be handled.
+message ProcessingResponse {
+  oneof response {
+    option (validate.required) = true;
+
+    // The processed ReadData containing potentially modified data for the request path.
+    // This should be sent in response to a ProcessingRequest with read_data.
+    ReadData read_data = 1;
```

Comment:
Do these need to be populated if the server does not actually modify the data?

Reply:
Yes, the data in the ProcessingRequest will be replaced as of now in Envoy's data plane. But I can foresee there are other use cases, like in observability mode that the response data will not be used. That part can be added in future IMHO.

Reply:
It seems like it will waste a lot of bandwidth for the server to send back the data if it doesn't actually need to modify it.  Doesn't the client keep that data buffered anyway?  If so, we should define the protocol such that the server only has to send back data if it wants to modify the data.

I don't think observability mode is relevant here, since I assume the ext_proc server wouldn't send back *any* response in that case.

Reply:
Synced up offline, and we can further add a knob in the ProcessingRequest to let the server decide if it needs send the buffered data back if it is unmodified, this can be a combination of filter knob and the processingRequest that can be added in future.